File: src/Exporter.NX.Translate/DocumentExporter.cs
// SPDX-License-Identifier: GPL-2.0-only
using System;
using System.Collections.Generic;
using System.Runtime.CompilerServices;
using Oblikovati.Exporter.NX.Model;
using Oblikovati.Exporter.NX.Recipe;

namespace Oblikovati.Exporter.NX.Translate
{
    /// <summary>One translated document and the file name it should be written to.</summary>
    public sealed class TranslatedDocument
    {
        public TranslatedDocument(string fileName, OblikovatiDocument document)
        {
            FileName = fileName;
            Document = document;
        }

        public string FileName { get; }

        public OblikovatiDocument Document { get; }
    }

    /// <summary>
    /// Walks a document tree and produces one <see cref="TranslatedDocument"/> per document:
    /// a part becomes one <c>.opd</c>; an assembly becomes one <c>.oad</c> plus the
    /// component documents it references (recursively). Components shared by several
    /// occurrences are exported once (deduped by reference), and the assembly's occurrences
    /// reference them by owner-relative file name, matching Oblikovati's reference graph.
    /// </summary>
    public sealed class DocumentExporter
    {
        private readonly DocumentTranslator _translator;

        public DocumentExporter(DocumentTranslator translator)
        {
            _translator = translator ?? throw new ArgumentNullException(nameof(translator));
        }

        public IReadOnlyList<TranslatedDocument> Export(NxDocument root, ExportReport report)
        {
            if (root == null) throw new ArgumentNullException(nameof(root));
            if (report == null) throw new ArgumentNullException(nameof(report));

            var files = new List<TranslatedDocument>();
            var fileNames = new Dictionary<NxDocument, string>(ReferenceComparer.Instance);
            var usedNames = new HashSet<string>(StringComparer.OrdinalIgnoreCase);
            Emit(root, files, fileNames, usedNames, report);
            return files;
        }

        // Returns the file name the document was (or will be) written to, emitting it once.
        private string Emit(
            NxDocument doc,
            List<TranslatedDocument> files,
            Dictionary<NxDocument, string> fileNames,
            HashSet<string> usedNames,
            ExportReport report)
        {
            if (fileNames.TryGetValue(doc, out string? existing))
            {
                return existing;
            }

            string fileName = UniqueName(doc, usedNames);
            fileNames[doc] = fileName;

            if (doc.Kind == NxDocumentKind.Assembly)
            {
                var occurrences = new List<OccurrenceData>();
                foreach (NxOccurrence occurrence in doc.Occurrences)
                {
                    string childFile = Emit(occurrence.Component, files, fileNames, usedNames, report);
                    occurrences.Add(OccurrenceTranslator.Translate(occurrence, childFile));
                }

                files.Add(new TranslatedDocument(fileName, _translator.TranslateAssembly(doc, occurrences)));
            }
            else
            {
                files.Add(new TranslatedDocument(fileName, _translator.Translate(doc, report)));
            }

            return fileName;
        }

        private static string UniqueName(NxDocument doc, HashSet<string> used)
        {
            string ext = doc.Kind == NxDocumentKind.Assembly ? ".oad" : ".opd";
            string baseName = doc.DisplayName.Length == 0 ? "untitled" : doc.DisplayName;
            string candidate = baseName + ext;
            int n = 2;
            while (!used.Add(candidate))
            {
                candidate = baseName + "_" + n++ + ext;
            }

            return candidate;
        }

        // Identity comparer (netstandard2.0 lacks ReferenceEqualityComparer): two component
        // references are the same exported file only when they are the same object.
        private sealed class ReferenceComparer : IEqualityComparer<NxDocument>
        {
            public static readonly ReferenceComparer Instance = new ReferenceComparer();

            public bool Equals(NxDocument? x, NxDocument? y) => ReferenceEquals(x, y);

            public int GetHashCode(NxDocument obj) => RuntimeHelpers.GetHashCode(obj);
        }
    }
}

File: python/oblikovati_exporter_nx/translate/document_exporter.py
# SPDX-License-Identifier: GPL-2.0-only
"""Walks a document tree and produces one TranslatedDocument per document.

A part becomes one ``.opd``; an assembly becomes one ``.oad`` plus the component
documents it references (recursively). Components shared by several occurrences are
exported once (deduped by reference), and the assembly's occurrences reference them by
owner-relative file name, matching Oblikovati's reference graph.
"""
from __future__ import annotations

from dataclasses import dataclass
from typing import Dict, List, Set

from ..model.document import NxDocument, NxDocumentKind
from ..recipe.document import OblikovatiDocument
from . import occurrence_translator
from .document_translator import DocumentTranslator
from .report import ExportReport


@dataclass
class TranslatedDocument:
    """One translated document and the file name it should be written to."""

    file_name: str
    document: OblikovatiDocument


class DocumentExporter:
    def __init__(self, translator: DocumentTranslator) -> None:
        self._translator = translator

    def export(self, root: NxDocument, report: ExportReport) -> List[TranslatedDocument]:
        files: List[TranslatedDocument] = []
        file_names: Dict[int, str] = {}  # keyed by object identity (id())
        used_names: Set[str] = set()  # lowercased, for case-insensitive dedup
        self._emit(root, files, file_names, used_names, report)
        return files

    # Returns the file name the document was (or will be) written to, emitting it once.
    def _emit(
        self,
        doc: NxDocument,
        files: List[TranslatedDocument],
        file_names: Dict[int, str],
        used_names: Set[str],
        report: ExportReport,
    ) -> str:
        existing = file_names.get(id(doc))
        if existing is not None:
            return existing

        file_name = _unique_name(doc, used_names)
        file_names[id(doc)] = file_name

        if doc.kind == NxDocumentKind.ASSEMBLY:
            occurrences = []
            for occurrence in doc.occurrences:
                child_file = self._emit(occurrence.component, files, file_names, used_names, report)
                occurrences.append(occurrence_translator.translate(occurrence, child_file))
            files.append(
                TranslatedDocument(file_name, self._translator.translate_assembly(doc, occurrences))
            )
        else:
            files.append(TranslatedDocument(file_name, self._translator.translate(doc, report)))

        return file_name


def _unique_name(doc: NxDocument, used: Set[str]) -> str:
    ext = ".oad" if doc.kind == NxDocumentKind.ASSEMBLY else ".opd"
    base = doc.display_name or "untitled"
    candidate = base + ext
    n = 2
    while candidate.lower() in used:
        candidate = f"{base}_{n}{ext}"
        n += 1
    used.add(candidate.lower())
    return candidate
